### ISSUE TITLE
'object' constraint statements appear to be incorrect

### ISSUE THREAD

**Issue by richardszalay** [dotnet/docs#17948]
&gt; Furthermore object is now valid as an explicit constraint. Absence of a constraint is now equivalent to an object? constraint (instead of object), but (unlike object before) object? is not prohibited as an explicit constraint.

Based on my testing, both these statements are untrue.

```
public void DoSomething&lt;T&gt;() where T : object {}
public void DoSomething&lt;T&gt;() where T : object? {}
```

Both the above methods fail to compile with the error:

&gt; CS0702: Constraint cannot be special class 'object'

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 43b5a98c-2cef-304c-1bbe-0ec466e21e0c
* Version Independent ID: bf6a616b-aa55-1679-67c2-e0a36d470dbc
* Content: [Nullable reference types - specification - C# 8.0 specification proposals](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/nullable-reference-types-specification#feedback)
* Content Source: [proposals/csharp-8.0/nullable-reference-types-specification.md](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/nullable-reference-types-specification.md)
* Product: **dotnet-csharp**
* Technology: **csharp-spec**
* GitHub Login: @dotnet-bot
* Microsoft Alias: **wiwagn**

**Comment by BillWagner** [comment]
Ping @MadsTorgersen 

I think this was updated to be the `notnull` constraint. is that correct?

**Comment by MadsTorgersen** [comment]
@BillWagner correct. For a while during feature design we thought we were going to enable `object` as a constraint, but instead decided a new `notnull` constraint better described the intuition behind what was intended:

``` c#
#nullable enable

static void M<T>(T t) where T : notnull { }

M<string?>(null); // Warning on `M<string?>`
M<string>(null);  // Warning on `null`
M<int?>(null);    // Warning on `M<int?>
M<int>(null);     // Error on `null`
```

HTH,

Mads

**Comment by BillWagner** [comment]
Thanks for clarifying that Mads.

We'll update the specification as part of the standards update process.  In the meantime, the language reference on [generic constraints](https://docs.microsoft.com/dotnet/csharp/language-reference/keywords/where-generic-type-constraint) does reflect the implemented behavior.
